### DISCUSSION TITLE
How can I customize the filtering of completion proposals ?

### DISCUSSION THREAD

**Discussion by ylussaud** [eclipse/lsp4e#548] [Q&A]
Hi,

I tried to find an entry point to change the way the completion proposal are filtered after the client retrieve them from the server. But when typing more text the client doesn't ask for a new list of proposals so I don't have access to the new typed text to filter the list of proposals.

For instance here I typed `eAA` after the completion was triggered:

![image](https://user-images.githubusercontent.com/809511/224714352-983fb5f6-3903-4ef4-8811-23a258023c7d.png)

And thank you for the great work !

regards,
Yvan.

**Comment by angelozerr** [comment]
Do you set CompletionList.isIncomplete to true in your language server? If you set to true, I suggest that you set to false.

**Reply by ylussaud** [reply]
I changed my `List<CompletionItem>` to a `CompletionList` setting `CompletionList.isIncomplete` to `false` but `TextDocumentService.complete()` only get called once. I noticed the following warning each time I type a caracter:

```
mars 14, 2023 10:04:07 AM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleCancellation
AVERTISSEMENT: Unmatched cancel notification for request id 10
```

After checking the warning is also present with `CompletionList.isIncomplete` to `true`. Could this be the source of my problem ?

**Reply by angelozerr** [reply]
Ooops sorry I wanted to say that you should set CompletionList.isIncomplete to true  to recompute the completion on language server side on each key stroke.

**Reply by ylussaud** [reply]
I tried both and `TextDocumentService.complete()` only get called once.

**Comment by mickaelistria** [comment]
Note that rather than attempting to override the default filter/sort of proposals, improving the built-in one directly in LSP4E would be welcome.

**Reply by ylussaud** [reply]
Do you have an entry point in the code so I can have a look? My idea was to implement a camel case filter.

**Reply by mickaelistria** [reply]
I would suggest you start by adding test cases that cover the expectation in IncompleteCompletionTest , then you'll probably want to tweak the LSCompletionProposalComparator so camel-case matches are given a bigger weigh than others. From your example, it seems relatively easy: if we have the matches, you can just prefer to put first the proposal that as the longest continuous match of a substring.

**Reply by ylussaud** [reply]
I had a look at the code you pointed, and it's only called once when the completion is triggered. None of the new key strokes passe through this code (in my example above `eAA`). I'll try to path LSP4E to request the completion on each key stroke after the completion is triggered...

**Reply by mickaelistria** [reply]
For completions (and filtering and so on) to be recomputed by the LS on all key strokes, you'd better set `isIncomplete=true` to the returned completion list.

**Reply by cergicd** [reply]
Is it possible to overwrite the default sort of proposals through an extension point? For example I would like the proposals coming from an LS(that i have configured through the provided extension point) to always show at the first position.

**Reply by rubenporras** [reply]
At the moment it is not possible. For that I think we would need to start by adding a property in the extension point that allows to set a precedence between the available servers and use that to sort the results. 

But even then more work is needed, given that the communication between the client and the server is asynchronous (and the server which responses you might want to show first might be slower).

**Comment by FlorianKroiss** [comment]
Maybe as a follow up to the original question: Is there an intended way to filter the proposals on the client side?

Some background: We are using lsp4e for editing Python files. The language server that we use includes dunder methods, e.g., `__init_subclass__`, in the proposal list. These methods are technical details within Python and might confuse less experienced Python developers, which is why we would like to filter them from the proposal list by default.

An obvious solution is to implement such a filter within the Language Server, but I thought that this should also be possible on the client side.
